sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HivePartitionFilteringSuite.scala

```diff
@@ -62,6 +63,9 @@ class HivePartitionFilteringSuite(version: String)
     properties = Map.empty
   )
 
+  // Avoid repeatedly constructing multiple hive instances that do not use direct sql
+  private var disableDirectSqlClient: HiveClient = _
```

Comment:
hmm can you explain why this is useful? 

looks like by caching the client, we just need to create 1 instead of 3 clients now, for each separate run on a Hive version (2 initialization in the existing code + 1 initialization in the new test case `getPartitionsByFilter: ds=20170101 when $fallbackKey=true`, is that correct?

I'm also curious why the OOM didn't happen before, as we were already using the `init(false)` in the tests.

Reply:
First, the size of `METASPACE_SIZE` is 1g
Then `HiveClientVersions` needs to test "0.14", "1.0", "1.1", "1.2", "2.0", "2.1", "2.2", "2.3", "3.0", "3.1", 11 hive versions.
Each version is initialized 3 times with `IsolatedClientLoader`, and hive-related classes cannot be unloaded after ut ends.

If we create hive client several times (val client = init(false)), it will cause `OutOfMemoryError: Metaspace`
For example, this test
https://github.com/cxzl25/spark/commit/888ac1ef068ed88f4802079d8120cac11b46999c  
 https://github.com/cxzl25/spark/runs/4023095713

Reply:
so you mean the test was fine before when we creates 2 clients with each Hive version, but start to failing when creating 3 clients?

Reply:
Yes, if I create a client multiple times, it will be OOM due to the limitation of MaxMetaspaceSize=1g.

Reply:
Got it, thanks. One nit: could you rename `disableDirectSqlClient` to something like `clientWithoutDirectSql`? the former looks like a boolean flag :)